tasksCode/CountCodeLines.cpp

```diff
@@ -0,0 +1,299 @@
+                fileContents.replace(fileContents.find(commentBody), commentBody.length(), cTypeEmptyComment);
+                ++localCommentLinesCount;
+            }
+        }
+        return localCommentLinesCount;
+    }
+
+    static void RemoveAllComments(std::string & fileContents) {
+        fileContents.push_back('\n');
+        // Iterate through lines in the file content
+        std::stringstream contentStream(fileContents);
+        std::string contentLine;
+        while (std::getline(contentStream, contentLine)) {
+            // Remove each instance of /**/ along with leading newline sign if
+            // it is the only text in the line, otherwise remove only /**/
+            std::smatch matchedComment;
+            if(std::regex_search(contentLine.cbegin(), contentLine.cend(), matchedComment, cTypeCommentRegex)) {
+                if (std::regex_match(std::string(matchedComment.prefix()), blankStringRegex) &&
+                    std::regex_match(std::string(matchedComment.suffix()), blankStringRegex))
+                    fileContents.erase(fileContents.find(contentLine + "\n"), contentLine.length() + 1);
+                else
+                    fileContents.erase(fileContents.find(matchedComment.str()), matchedComment.length());
+            }
+        }
+        fileContents.erase(fileContents.length()-1, 1);
+    }
+
+    static std::pair<CounterType, CounterType> CountCodeAndBlankLines(std::string & fileContents)
+    {
+        CounterType localCodeLinesCount = 0;
+        CounterType localBlankLinesCount = 0;
+        // Iterate through each line in the file content and increase the corresponding counters
+        std::stringstream fileContentsStream(fileContents);
+        std::string contentLine;
+        while (std::getline(fileContentsStream, contentLine)) {
+            if (std::regex_match(contentLine, blankStringRegex))
+                ++localBlankLinesCount;
+            else
+                ++localCodeLinesCount;
+        }
+        return std::make_pair(localCodeLinesCount, localBlankLinesCount);
+    }
+
+    void JoinThreads()
+    {
+        // Join all threads to ensure that all files have been processed
+        std::for_each(fileProcesses.begin(), fileProcesses.end(),
+                      [](boost::thread & fileProcess) {
+                          if(fileProcess.joinable())
+                              fileProcess.join();
+                      });
+        fileProcesses.clear();
+    }
+public:
+    ConcurrentLineCounter() {
+        fileProcesses = std::vector<boost::thread>();
+        filesCount        = 0;
+        blankLinesCount   = 0;
+        codeLinesCount    = 0;
+        commentLinesCount = 0;
+    };
+
+    void operator()(const std::string & fileName)
+    {
+        // Count the lines in the file with a given filename
+        filesCount += 1;
+        fileProcesses.emplace_back(ConcurrentLineCounter::CountLines, fileName);
```

Comment:
It might be a bad idea to create a vector for each thread. When we create more threads than our operation system can support (more then the cores number) our physical thread will switch any time thread's function will work. It can be time consuming  to make this switch operation. I'd suggest to create only the number of threds which our system can support (FE 4) and run any task to this thread when it's don't have any work to do.

Reply:
You are right also because system creates for each thread its internal memory. I.e. if there are lot of files to be analyzed, the RAM is going to be overloaded.